cli/loader_lib.c

```diff
@@ -153,6 +154,82 @@ JL_DLLEXPORT const char * jl_get_libdir()
     return lib_dir;
 }
 
+#ifdef LIBSTDCXXPROBE
+// Write the path to the libstdcxx to load to "path", which should be a buffer of at least
+// JL_PATH_MAX, and the length of that string (excluding null terminator) to "written"
+// Returns 1 on success or 0 on error, in which case the caller should abort the process.
+#define JL_DEFAULT_STDCXXLIB "libstdc++.so.6"
+static int libstdcxxprobe(char* path, size_t* written) {
+    // On Linux, it can happen that the system as a newer libstdc++ than the one we ship,
+    // which can break loading of some system libraries:
+    // <https://github.com/JuliaLang/julia/issues/34276>.
+    int ret;
+    pid_t pid;
+    int fork_pipe[2];
+    void *handle;
+    char *dlerr;
+
+    // Create the pipe and child process.
+    ret = pipe(fork_pipe);
+    if (ret == -1) return 0;
+    pid = fork();
+    if (pid == -1) return 0;
+
+    if (pid == (pid_t) 0) { // Child process.
+        ret = close(fork_pipe[0]);
+        if (ret == -1) exit(1);
+
+        // Open the first available libstdc++.so
+        handle = dlopen("libstdc++.so", RTLD_LAZY);
+        if (!handle) exit(0);
+
+        // See if the version is compatible
+        dlerr = dlerror(); // clear out dlerror
+        void* sym = dlsym(handle, "GLIBCXX_3.4.29");
```

Comment:
It'd nice to have a way to keep this in sync with https://github.com/JuliaLang/julia/blob/9fd408723aab0974f399b8ffc952af4b4d20b6f7/deps/csl.mk#L25 although it's a bit tricky because that refers to the next possible versions of libstdc++

Reply:
We can take the first value, split it on periods, subtract one from the patch version and recombine. That’s doable in make

Reply:
Naively, I think you can do something like
```makefile
GLIBCXX_MAJOR := $(shell echo $(GLIBCXX_VERSION) | cut -d'.' -f1)
GLIBCXX_MINOR := $(shell echo $(GLIBCXX_VERSION) | cut -d'.' -f2)
GLIBCXX_PATCH := $(shell echo $(GLIBCXX_VERSION) | cut -d'.' -f3)
THING := $(GLIBCXX_MAJOR).$(GLIBCXX_MINOR).$(shell expr $(GLIBCXX_PATCH) - 1)
```

Reply:
Actually, I'm wondering if here we should actually search for the next version of GLIBCXX, instead of the current one, so that we can simply grab the first item of `CSL_NEXT_GLIBCXX_VERSION`.

Reply:
Agreed; since we are not shipping the absolute latest libstdc++, we know that the first element of `CSL_NEXT_GLIBCXX_VERSION` exists, and is a good indication that we should use the system libstdc++.

Reply:
Two remarks:
1. Isn't that `CSL_NEXT_GLIBCXX_VERSION` pattern incomplete? It doesn't match e.g. `3.6.0`... or are you really only interested in the literal *next* version? If not, a pattern like `GLIBCXX_3\.[5-9]\.` would be better, but of course it'd also not deal with `3.10.0`, so then you'd also need `GLIBCXX_3\.[1-9][0-9]+\.` or so (untested)
2. how about doing the reverse, and storing the value `3.4.29` (or even the three numbers 3, 4, 29 in the Makefile, and building `CSL_NEXT_GLIBCXX_VERSION` from that? That seems simpler.